docs/docs/config-reference.md
---
id: config-reference
title: Configuration Reference
---

Kodiak's configuration file is a TOML file and should be placed at `.kodiak.toml` (repository root) or `.github/.kodiak.toml`.

If Kodiak cannot find configuration in your repository, Kodiak will look in your `.github` repository. Kodiak must be installed on your `.github` repository to read from it.

For Kodiak to run on a pull request:

1. Kodiak must be installed on the repository.
2. A configuration file must exist in the repository.
3. GitHub branch protection must exist on the target branch.

## configuration fields

### `version`

- **type:** `number`
- **required:** `true`

`1` is the only valid setting for this field.

### `merge.automerge_label`

- **type:** `string` or `string[]`
- **default:** `"automerge"`

Label to enable Kodiak to merge a PR.

By default, Kodiak will only act on PRs that have this label. You can disable this requirement via `merge.require_automerge_label`.

Kodiak can only merge pull requests passing your GitHub branch protection rules.

```toml
merge.automerge_label = "🚀 merge it!"
```

If multiple labels are specified in an array, any of the specified labels will trigger merge.

```toml
merge.automerge_label = ["🚢 it!", "merge it!"]
```

### `merge.require_automerge_label`

- **type:** `boolean`
- **default:** `true`

Require that the automerge label (`merge.automerge_label`) be set for Kodiak to merge a PR.

When disabled, Kodiak will immediately attempt to merge any PR that passes all GitHub branch protection requirements.

### `merge.show_missing_automerge_label_message`

- **type:** `boolean`
- **default:** `true`

Show a status message when automerge label is required but missing on a PR.

When disabled, no status message is shown until the automerge label has been added to the PR.

### `merge.automerge_dependencies.versions`

- **type:** `string[]`
- **options:** `"major"`, `"minor"`, `"patch"`
- **default:** `[]`

Kodiak will only automerge version upgrade types in this list. The author of the pull request must also be listed in [`merge.automerge_dependencies.usernames`](#mergeautomerge_dependenciesusernames).

Like [`merge.automerge_label`](#mergeautomerge_label), your pull request can only be merged if it passes your GitHub branch protection rules.

See ["Configuring automerge by upgrade type"](recipes.md##configuring-automerge-by-upgrade-type) for a full example.

```toml
# .kodiak.toml
[merge.automerge_dependencies]
# only auto merge "minor" and "patch" version upgrades.
# do not automerge "major" version upgrades.
versions = ["minor", "patch"]
usernames = ["dependabot", "renovate"]
```

Dependency upgrade types are parsed from the pull request title and body. The following table shows version upgrade examples for Dependabot pull request titles:

| title                           | upgrade |
| ------------------------------- | ------- |
| Bump lodash from 1.0.0 to 1.0.1 | patch   |
| Bump lodash from 2.5.1 to 2.8.0 | minor   |
| Bump lodash from 4.2.1 to 5.0.0 | major   |

If Kodiak cannot determine the upgrade type from the pull request title, Kodiak will not automerge the pull request.

See the [tests file](https://github.com/chdsbd/kodiak/blob/b1893ee6add4a1533bdac77999aad698e0b2e74c/bot/kodiak/test_dependencies.py#L10-L35) for more examples.

### `merge.automerge_dependencies.usernames`

- **type:** `string[]`
- **default:** `[]`

Kodiak will only automerge dependency upgrades for pull request authors in this list.

See ["Configuring automerge by upgrade type"](recipes.md##configuring-automerge-by-upgrade-type) for a full example.

```toml
# .kodiak.toml
[merge.automerge_dependencies]
versions = ["minor", "patch"]
# automerge by upgrade version for pull requests authored by dependabot or renovate.
usernames = ["dependabot", "renovate"]
```

### `merge.block_on_neutral_required_check_runs`

- **type:** `boolean`
- **default:** `false`

GitHub considers a "neutral" check run conclusion as passing GitHub branch protection requirements.

When enabled, pull requests won't be merged if a required check run has a neutral conclusion.

<span id="mergeblacklist_title_regex"/> <!-- handle old links -->

### `merge.blocking_title_regex`

- **type:** `string`
- **default:** `"^WIP:.*"`
- **options:** Regex pattern or `""`
- **alias:** `merge.blacklist_title_regex`

If a PR's title matches this regex, Kodiak will not merge the PR. This is useful
to prevent merging work-in-progress PRs.

Setting `merge.blocking_title_regex = ""` disables this option.

#### example

```
merge.blocking_title_regex = ".*DONT\s*MERGE.*"
```

> **NOTE:** `merge.blocking_title_regex` is a new name for `merge.blacklist_title_regex`. Both options behave identically.

<span id="mergeblacklist_labels"/> <!-- handle old links -->

### `merge.blocking_labels`

- **type:** `string[]`
- **default:** `[]`
- **options:** List of label names
- **alias:** `merge.blacklist_labels`

Kodiak will not merge a PR with any of these labels.

#### example

```
merge.blocking_labels = ["wip"]
```

> **NOTE:** `merge.blocking_labels` is a new name for `merge.blacklist_labels`. Both options behave identically.

### `merge.method`

- **type:** `string`
- **default:** first valid merge method in list `"merge"`, `"squash"`, `"rebase"`
- **options:** `"merge"`, `"squash"`, `"rebase"`, `"rebase_fast_forward"`

Choose merge method for Kodiak to use.

If not configured, the first valid merge method in the list `"merge"`, `"squash"`, `"rebase"` will be used. For example, if `"squash"` and `"rebase"` are the only valid options on a repository, Kodiak will choose `"squash"` since it's the first valid option.

Kodiak will report a configuration error if the selected merge method is disabled for a repository.

If you're using the "Require signed commits" GitHub Branch Protection setting to require commit signatures, _`"merge"` or `"squash"` are the only compatible options_. `"rebase"` will cause Kodiak to raise a configuration error.

The `"rebase_fast_forward"` option is similar to `"rebase"`, but commits are rebased without being rewritten. This option is helpful if your build system depends on git commit hashes instead of git tree hashes for build deduplication.

Since the `"rebase_fast_forward"` option only works with fast forward merges, you must enable the "Require branches to be up to date before merging" branch protection setting with at least one status check.

### `merge.delete_branch_on_merge`

- **type:** `boolean`
- **default:** `false`

Once a PR is merged, delete the branch.

This option behaves like the GitHub repository setting "Automatically delete head branches", which automatically deletes head branches after pull requests are merged.

If "Automatically delete head branches" is enabled on the repository via the GitHub UI, Kodiak will _not_ attempt branch deletion.

### `merge.block_on_reviews_requested`

- **type:** `boolean`
- **default:** `false`

> **DEPRECATED**
>
> Due to limitations with the GitHub API this feature is fundamentally broken and cannot be fixed. Prefer the GitHub branch protection "required reviewers" setting instead.
>
> When a user leaves a comment on a PR, GitHub counts that as satisfying a review request, so the PR will be allowed to merge, even though a reviewer was likely just starting a review.
>
> See this issue comment for more information: [chdsbd/kodiak#153 (comment)](https://github.com/chdsbd/kodiak/issues/153#issuecomment-523057332)

If you request review from a user, don't merge until that user provides a review, even if the PR is passing all status checks.

### `merge.notify_on_conflict`

- **type:** `boolean`
- **default:** `true`

If there is a merge conflict, make a comment on the PR and remove the
automerge label.

This option only applies when `merge.require_automerge_label` is enabled.

### `merge.optimistic_updates`

- **type:** `boolean`
- **default:** `true`

_Assuming your status checks pass most of the time, it's useful to leave `merge.optimistic_updates` enabled._

If Kodiak is merging an out-of-date pull request that has running status checks, update the pull request's branch without waiting for the running status checks to finish.

This setting can speed up merges by not waiting for status checks to finish on out-of-date pull requests.

> **Note:** The "Require branches to be up to date before merging" GitHub Branch Protection setting must be enabled for Kodiak to update branches.

### `merge.dont_wait_on_status_checks`

- **type:** `string[]`
- **default:** `[]`
- **options:** List of check names

Don't wait for specified status checks when merging a PR. If a configured status check is incomplete when a PR is being merged, Kodiak will skip the PR.

Use this option for status checks that run indefinitely, like deploy jobs or the WIP GitHub App.

#### example

```
merge.dont_wait_on_status_checks = ["ci/circleci: deploy", "WIP"]
```

### `merge.update_branch_immediately`

- **type:** `boolean`
- **default:** `false`

> **DEPRECATED**
>
> Prefer [`update.always`](#updatealways), which will deliver better behavior in most use cases. `merge.update_branch_immediately` only affects PRs eligible for merging, while `update.always` will keep all PRs up-to-date.

Update PRs that are passing all branch requirements or are waiting for status checks to pass.

### `merge.prioritize_ready_to_merge`

- **type:** `boolean`
- **default:** `false`

If a PR is passing all checks and is able to be merged, merge it without
placing it in the merge queue.

This option adds some unfairness where PRs waiting in the queue the longest are not served first.

### `merge.priority_merge_label`

- **type:** `string`
- **default:** `null`

When the label is applied to a pull request, the pull request will be placed at the front of the merge queue.

If two PRs have the `merge.priority_merge_label` label, either one may be first in queue.

### `merge.do_not_merge`

- **type:** `boolean`
- **default:** `false`

Never merge a PR. This option can be used with `update.always` to automatically update a PR without merging.

### `merge.message.title`

- **type:** `enum`
- **default:** `"github_default"`
- **options:** `"github_default"`, `"pull_request_title"`

By default (`"github_default"`), the title depends on [merge.method](config-reference.md#mergemethod):

- In case of `"merge"`, GitHub will create a merge commit title of the form, `Merge pull request #X from Y/Z`.
- In case of `"rebase"`, there is no merge commit, and all commits of your pull request will be added to the target unchanged.
- In case of `"squash"`, the squashed commit title depends on the number of commits in the pull request. If the PR contains only one commit, GitHub uses the title of that commit for the squashed commit title. If there are multiple commits, GitHub uses the pull request title. In both cases the PR number is appended to the title, like `(#543)`. [See the official GitHub documentation for more information](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#merge-message-for-a-squash-merge).

For `"pull_request_title"`, Kodiak uses the pull request title for both merge and squash commits.

### `merge.message.body`

- **type:** `enum`
- **default:** `"github_default"`
- **options:** `"github_default"`, `"pull_request_body"`, `"empty"`

By default (`"github_default"`), GitHub combines the titles of a PR's commits to create the body
text of a merge commit. `"pull_request_body"` uses the content of the PR to generate
the body content while `"empty"` sets an empty body.

### `merge.message.include_pr_number`

- **type:** `boolean`
- **default:** `true`

Add the PR number to the merge commit title.

This setting replicates GitHub's behavior of automatically adding the PR number to the title of merges created through the UI.

This option only applies when `merge.message.title` does not equal `"github_default"`.

### `merge.message.body_type`

- **type:** `enum`
- **default:** `"markdown"`
- **options:** `"markdown"`, `"plain_text"`, `"html"`

Control the text used in the merge commit. The GitHub default is markdown, but `"plain_text"` or `"html"` can be used to render the pull request body as text or HTML.

This option only applies when `merge.message.body = "pull_request_body"`.

### `merge.message.strip_html_comments`

- **type:** `boolean`
- **default:** `false`

Strip HTML comments (`<!-- some HTML comment -->`) from merge commit body.

This setting is useful for stripping HTML comments created by PR templates.

This option only applies when `merge.message.body_type = "markdown"`.

### `merge.message.cut_body_before`

- **type:** `string`
- **default:** `null`

Remove all content _before_ the configured string in the pull request body.

This setting is useful when we want to include only a part of the pull request description as the commit message.

This option only applies when `merge.message.body_type = "markdown"`.

### `merge.message.cut_body_after`

- **type:** `string`
- **default:** `null`

Remove all content _after_ the configured string in the pull request body.

This setting is useful when we want to include only a part of the pull request description as the commit message.

This option only applies when `merge.message.body_type = "markdown"`.

### `merge.message.cut_body_and_text`

- **type:** `boolean`
- **default:** `False`

Remove the string used to handle body in `cut_body_before` and `cut_body_after`.

This setting is useful if you believe that the separator doesn't belong in the commit message.

This option only applies if `cut_body_after` or `cut_body_before` is in use.

### `merge.message.include_coauthors`

- **type:** `boolean`
- **default:** `false`

If a PR includes commits authored by other users, add those users as coauthors to the pull request.

Coauthors are added using the [`Co-authored-by` trailer syntax](https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors#creating-co-authored-commits-on-github).

This setting only applies when `merge.message.body = "pull_request"` is set.

#### Example commit

Notice the `Co-authored-by` trailer added by this option.

```text
Add new github login flow (#17)

This change adds the new GitHub social login flow. Steve provided some great UI tweaks.

Co-authored-by: Steve Dignam <7340772+sbdchd@users.noreply.github.com>
```

<img height="80px" title="Coauthors example screenshot" src="/img/coauthors-example.png"/>

### `merge.message.include_pull_request_author`

- **type:** `boolean`
- **default:** `false`

Add the pull request author as a coauthor of the merge commit using `Co-authored-by: jdoe <828352+jdoe@users.noreply.github.com>` syntax.

This setting only applies when `merge.message.body = "pull_request"` is set.

This setting was added to mitigate the fallout of GitHub's change to the
squash method on March 4th, 2020. GitHub reverted their change around
March 6th, 2020, making this option no longer necessary.

### `merge.message.include_pull_request_url`

- **type:** `boolean`
- **default:** `false`

Append the pull request's URL to the commit message body, prefixed with `PR-URL:`.

This can make accessing the relevant PR for a given commit easier.

This setting only applies when `merge.message.body = "pull_request"` is set.

#### Example commit

Notice the `PR-URL` trailer added by this option.

```text
Add new github login flow (#17)

This change adds the new GitHub social login flow.

PR-URL: https://github.com/ghost/app/pull/17
```

### `update.always`

- **type:** `boolean`
- **default:** `false`

Update a PR whenever out of date with the base branch. The PR will be
updated regardless of merge requirements (e.g. failing status
checks, missing reviews, blacklist labels).

Kodiak will only update PRs with the `merge.automerge_label` label or if `update.require_automerge_label = false`.

When enabled, _Kodiak will not be able to efficiently update PRs._ If you have multiple PRs against a target like `master`, any time a commit
is added to `master` _all_ of those PRs against `master` will update. For `N` PRs against a target you will see at least `N(N-1)/2` updates. If
this configuration option was disabled you would only see at least `N-1` updates.

### `update.require_automerge_label`

- **type:** `boolean`
- **default:** `true`

When enabled, Kodiak will only update PRs that have an automerge label (configured via `merge.automerge_label`).

When disabled, Kodiak will update any PR.

This option only applies when `update.always = true`.

### `update.autoupdate_label`

- **type:** `string`
- **default:** `null`

Pull requests with the `update.autoupdate_label` will be updated when they are out-of-date with their base branch.

This configuration option is similar to `update.always`, but only pull requests with the configured label are affected.

#### example

```toml
version = 1

[merge]
automerge_label = "ship it!"

[update]
autoupdate_label = "update me please!"
```

<span id="updateblacklist_usernames"/> <!-- handle old links -->

### `update.ignored_usernames`

- **type:** `string[]`
- **default:** `[]`
- **alias:** `update.blacklist_usernames`

When a pull request's author matches `update.ignored_usernames`, Kodiak will never update the pull request, unless `update.autoupdate_label` is applied to the pull request.

If the user is a bot user, remove the `[bot]` suffix from their username. So instead of `dependabot-preview[bot]`, use `dependabot-preview`.

#### example

```
update.ignored_usernames = ["bernard-lowe", "dependabot-preview"]
```

> **NOTE:** `update.ignored_usernames` is a new name for `update.blacklist_usernames`. Both options behave identically.

### `approve.auto_approve_usernames`

- **type:** `string[]`
- **default:** `[]`
- **options:** List of GitHub usernames

If a PR is opened by a user with a username in the `approve.auto_approve_usernames` list, Kodiak will automatically add an approval to the PR.

If Kodiak's review is dismissed, it will add a review again.

This setting is useful when the "Required approving reviews" GitHub Branch Protection setting is configured and dependency upgrade bots like dependabot, greenkeeper, etc, run on the repository. When these bots open a PR, Kodiak can automatically add a review so dependency upgrade PRs can be automatically merged.

See the "[Automated dependency updates with Dependabot](recipes.md#automated-dependency-updates-with-dependabot)" recipe for an example of this feature in action.

### `approve.auto_approve_labels`

- **type:** `string[]`
- **default:** `[]`
- **options:** List of label names

If a PR has a label assigned which is in the `approve.auto_approve_labels` list, Kodiak will automatically add an approval to the PR.

If Kodiak's review is dismissed, it will add a review again.

This setting is useful when the "Required approving reviews" GitHub Branch Protection setting is configured and dependency upgrade bots like dependabot, greenkeeper, etc, run on the repository. When these bots open a PR with the specified label(s), Kodiak can automatically add a review so dependency upgrade PRs can be automatically merged.

## configuration labels

Some configuration options set in `.kodiak.toml` can be overridden with pull request labels.

Please [open an issue on GitHub](https://github.com/chdsbd/kodiak/issues/new/choose) if you would like support for more options.

### `merge.method`

The [`merge.method`](config-reference.md#mergemethod) option can be overridden with a label, allowing you to merge a pull request with a `merge.method` that differs from your `.kodiak.toml` setting.

#### Example

Label your pull request with the `kodiak: merge.method = 'rebase'` label to set `merge.method` to "rebase" for your pull request.

Here's a table of all the options:

| merge method | label                             |
| ------------ | --------------------------------- |
| merge        | `kodiak: merge.method = 'merge'`  |
| squash       | `kodiak: merge.method = 'squash'` |
| rebase       | `kodiak: merge.method = 'rebase'` |

## full examples

### minimal

```toml
# .kodiak.toml
# docs: https://kodiakhq.com/docs/config-reference
version = 1
```

### all options

Below is a Kodiak config with all options set and commented.

```toml
# .kodiak.toml

# Kodiak's configuration file should be placed at `.kodiak.toml` (repository
# root) or `.github/.kodiak.toml`.
# docs: https://kodiakhq.com/docs/config-reference

# version is the only required setting in a kodiak config.
# `1` is the only valid setting for this field.
version = 1

[merge]
# Label to enable Kodiak to merge a PR.

# By default, Kodiak will only act on PRs that have this label. You can disable
# this requirement via `merge.require_automerge_label`.
automerge_label = "automerge" # default: "automerge"

# Require that the automerge label (`merge.automerge_label`) be set for Kodiak
# to merge a PR.
#
# When disabled, Kodiak will immediately attempt to merge any PR that passes all
# GitHub branch protection requirements.
require_automerge_label = true

# If a PR's title matches this regex, Kodiak will not merge the PR. This is
# useful to prevent merging work-in-progress PRs.
#
# Setting `merge.blocking_title_regex = ""` disables this option.
blocking_title_regex = "" # default: "^WIP:.*", options: "" (disables regex), a regex string (e.g. ".*DONT\s*MERGE.*")

# Kodiak will not merge a PR with any of these labels.
blocking_labels = [] # default: [], options: list of label names (e.g. ["wip"])

# Choose merge method for Kodiak to use.
#
# Kodiak will report a configuration error if the selected merge method is
# disabled for a repository.
#
# If you're using the "Require signed commits" GitHub Branch Protection setting
# to require commit signatures, "merge" or "squash" are the only compatible options. "rebase" will cause Kodiak to raise a configuration error.
method = "merge" # default: first valid merge method in list "merge", "squash", "rebase", options: "merge", "squash", "rebase"

# Once a PR is merged, delete the branch. This option behaves like the GitHub
# repository setting "Automatically delete head branches", which automatically
# deletes head branches after pull requests are merged.
delete_branch_on_merge = false # default: false

# DEPRECATED
#
# Due to limitations with the GitHub API this feature is fundamentally broken
# and cannot be fixed. Prefer the GitHub branch protection "required reviewers"
# setting instead.
#
# When a user leaves a comment on a PR, GitHub counts that as satisfying a
# review request, so the PR will be allowed to merge, even though a reviewer was
# likely just starting a review.
#
# See this issue comment for more information:
# https://github.com/chdsbd/kodiak/issues/153#issuecomment-523057332
#
# If you request review from a user, don't merge until that user provides a
# review, even if the PR is passing all status checks.
block_on_reviews_requested = false # default: false

# If there is a merge conflict, make a comment on the PR and remove the
# automerge label. This option only applies when `merge.require_automerge_label`
# is enabled.
notify_on_conflict = true # default: true

# Don't wait for in-progress status checks on a PR to finish before updating the
# branch.
optimistic_updates = false # default: true


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Added support for GitHub's merge queue. When a merge queue is required for the base branch, Kodiak adds the pull request to the merge queue instead of merging it. (#927)

### Fixed

- Handle a redacted branch protection push allowance actor. (#926)

## 0.60.0 - 2026-05-02

### Fixed

- Don't warn about "missing push allowance" when a ruleset's bypass actor is redacted by the GitHub API. (#917)

## 0.59.1 - 2026-03-12

### Fixed

- Suppress warning about null branch protection rules. (#912)

## 0.59.0 - 2026-03-12

### Changed

- Reduced GitHub GraphQL API cost from 4 units to 2 units. (#911)

## 0.58.1 - 2026-03-12

### Fixed

- Handle bots as requested reviewers. (#911)

## 0.58.0 - 2026-03-12

### Changed

- Removed workaround for GitHub GraphQL bug introduced in #908. (#910)

## 0.57.0 - 2026-02-26

### Added

- Added timeout for processing webhook ingest events. (#909)

### Fixed

- Added workaround for GitHub GraphQL bug. (#908)

### Changed

- Split GitHub API throttling into separate REST API and GraphQL API throttlers. (#908)

## 0.56.0 - 2025-12-12

### Added

- Added support for specifying Redis username. (#881) Thanks @androa!
- Added support for GitHub rulesets, in addition to existing branch protection rule support. (#905) Thanks @qefir!

### Changed

- Migrated from poetry to uv and updated Python from 3.7 to 3.8. (#900)
- Replaced flake8, black and pylint with ruff. (#903)
- Switched from CircleCI to GitHub Actions. (#888)

## 0.55.0 - 2023-02-04

### Added

- Parse pull request body to find dependabot versions. (#845)

### Changed

- Replaced `asyncio-redis` with `redis-py` asyncio to improve network reliability. (#867)

### Fixed

- Delete subscriptions from web API db on cancellation. (#838)
- Add workaround for bug in GitHub GraphQL API. (#842)
- Retry on merge being blocked to workaround GitHub API consistency issues. (#846)
- Fix schema to handle null requestedReviewer. (#847)
- Add better support for required codeowner reviews. (#849)
- Fix HTTP performance with `httpx` by reusing the ssl context. (#852)

## 0.54.0 - 2022-07-04

### Fixed

- Fix parsing a PR without an author / without commits / with null CheckRuns (#828, #831, #827)
- internal: remove FastAPI dependency and use Starlette (#826)

## 0.53.0 - 2022-07-03

### Fixed

- Support m1 macos for development. (#824)

## 0.52.0 - 2022-06-06

### Added

- Added `approve.auto_approve_labels` configuration option to auto approve PRs (#817). Thanks @paescuj!

## 0.51.0 - 2022-04-24

### Fixed

- Fixed `merge.message.strip_html_comments` to correctly handle non-ascii pull request bodies. (#805)

## 0.50.0 - 2022-02-05

### Added

- Added `merge.block_on_neutral_required_check_runs` option to stop Kodiak from merging a pull request if a require check run has a neutral conclusion. (#785)

- Added `merge.message.cut_body_and_text` to remove separator when using
  `cut_body_after` or `cut_body_before` (#771)

## 0.49.0 - 2021-11-26

### Fixed

- Improved robustness for handling pull request required reviews. (#762, #763, #765)

## 0.48.0 - 2021-11-23

### Added

- Added `merge.show_missing_automerge_label_message` option to toggle missing automerge label message. Thanks @paescuj! (#757)

## 0.47.1 - 2021-11-17

### Fixed

- Fix Redis password auth support. Thanks @etiennetremel (#754)

## 0.47.0 - 2021-11-10

### Fixed

- Improved webhook and merge queue latencies by reducing concurrent API requests made by raw webhook ingestion. (#744)

## 0.46.0 - 2021-09-24

### Added

- `.github/` repository support for configuration. Create a repository named `.github` with your `.kodiak.toml` to centralize your `.kodiak.toml` configuration. Thanks @LeoQuote! (#729)

### Changed

- `merge.message.include_pull_request_url` to prefix URL with `PR-URL:`. Thanks @jbergstroem! (#730)

### Fixed

- support GitHub Enterprise 3.0. (#734)

## 0.45.0 - 2021-08-23

### Added

- Support Renovate for dependency updates by upgrade type. (#713, #715)

## 0.44.0 - 2021-07-20

### Added

- TLS support for Redis connections. Thanks @zetaron! (#695)
- requiresConversationResolution branch protection support. (#692)

## 0.43.0 - 2021-07-10

### Fixed

- fixed Kodiak incorrectly saying a PR was not mergeable when there were duplicated check runs. (#688)

## 0.42.1 - 2021-06-23

### Fixed

- fixed Kodiak-breaking bug from release 0.42.0 (#676)
- fixed short HTTP timeouts for GitHub API requests negatively impacting Kodiak's reliability. (#678)

## 0.42.0 - 2021-06-20

### Added

- use [fixed IP addresses](https://kodiakhq.com/docs/troubleshooting#github-enterprise-ip-allow-list) for sending GitHub API requests to support GitHub Enterprise Cloud IP address allow list. (#666, #673)

### Fixed

- fixed typo in config reference documentation. Thanks @AndersDJohnson! (#663)

## 0.41.0 - 2021-06-04

### Changed

- Configuration file (`.kodiak.toml`) is now read from a pull request's base branch instead of repository default branch. (#152, #634)

### Fixed

- fixed handling of "check run" webhooks to ignore pull requests from other repositories. (#644)
- When fetching open pull requests, Kodiak now loads up to 20 pages of 100 pull requests. Previously Kodiak would only fetch a single page of 30 pull requests. Thanks [@wuub](https://github.com/wuub) for the contribution! (#649)

## 0.40.0 - 2021-03-26

### Changed

- requested the repository "Workflow" permission to allow Kodiak to merge pull requests that modify GitHub Workflow files. (#641, #642)

## 0.39.0 - 2021-03-21

### Fixed

- fix merge queues not displaying in dashboard and merging pull request not appearing. (#632, #623, #621, #635)

bot/kodiak/config.py
from __future__ import annotations

from enum import Enum
from typing import List, Optional, Union

import toml
from pydantic import BaseModel, ValidationError, validator
from typing_extensions import Literal


class MergeMethod(str, Enum):
    merge = "merge"
    squash = "squash"
    rebase = "rebase"
    rebase_fast_forward = "rebase_fast_forward"


class MergeTitleStyle(Enum):
    github_default = "github_default"
    pull_request_title = "pull_request_title"


class MergeBodyStyle(Enum):
    github_default = "github_default"
    pull_request_body = "pull_request_body"
    empty = "empty"


class BodyText(Enum):
    plain_text = "plain_text"
    markdown = "markdown"
    html = "html"


class MergeMessage(BaseModel):
    """
    https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button
    """

    title: MergeTitleStyle = MergeTitleStyle.github_default
    body: MergeBodyStyle = MergeBodyStyle.github_default
    include_pr_number: bool = True
    body_type: BodyText = BodyText.markdown
    strip_html_comments: bool = False
    include_pull_request_author: bool = False
    include_coauthors: bool = False
    include_pull_request_url: bool = False
    cut_body_before: str = ""
    cut_body_after: str = ""
    cut_body_and_text: bool = False


# this pattern indicates that the user has the field unset.
UNSET_TITLE_REGEX = ":::|||kodiak|||internal|||reserved|||:::"
DEFAULT_TITLE_REGEX = "^WIP:.*"


class AutomergeDependencies(BaseModel):
    versions: List[Literal["major", "minor", "patch"]] = []
    usernames: List[str] = []


class Merge(BaseModel):
    # label or labels to enable merging of pull request.
    automerge_label: Union[str, List[str]] = "automerge"
    automerge_dependencies: AutomergeDependencies = AutomergeDependencies()
    # if disabled, kodiak won't require a label to queue a PR for merge
    require_automerge_label: bool = True
    # Show message when automerge label is missing on a PR
    show_missing_automerge_label_message: bool = True
    # regex to match against title and block merging. Set to empty string to
    # disable check.
    blacklist_title_regex: str = (
        UNSET_TITLE_REGEX  # deprecated for blocking_title_regex
    )
    blocking_title_regex: str = UNSET_TITLE_REGEX
    # labels to block merging of pull request
    blacklist_labels: List[str] = []  # deprecated for blocking_labels
    blocking_labels: List[str] = []
    # action to take when attempting to merge PR. An error will occur if method
    # is disabled for repository
    method: Optional[MergeMethod] = None
    # delete branch when PR is merged
    delete_branch_on_merge: bool = False
    # DEPRECATED
    # this feature is flawed and cannot be fixed. see
    # https://github.com/chdsbd/kodiak/issues/153#issuecomment-523057332.
    #
    # block merging if there are outstanding review requests
    block_on_reviews_requested: bool = False
    # block merging if there are neutral required check runs.
    block_on_neutral_required_check_runs: bool = False
    # comment on merge conflict and remove automerge label
    notify_on_conflict: bool = True
    # don't wait for status checks to run before updating branch
    optimistic_updates: bool = True
    # configuration for commit message of merge
    message: MergeMessage = MergeMessage()
    # status checks that we don't want to wait to complete when they are in a
    # pending state. This is useful for checks that will complete in an
    # indefinite amount of time, like the wip-app checks or status checks
    # requiring manual approval.
    dont_wait_on_status_checks: List[str] = []
    # DEPRECATED
    # This setting only updates PRs that are passing passing all
    # requirements or waiting for status checks to pass. `update.always = True`
    # will deliver better behavior in many use cases.
    #
    # immediately update a PR whenever the target updates
    update_branch_immediately: bool = False
    # if a PR is passing all checks and is able to be merged, merge it without
    # placing it in the queue. This will introduce some unfairness where those
    # waiting in the queue the longest will not be served first.
    prioritize_ready_to_merge: bool = False
    # when applied to a PR, add the PR to the front of the merge queue.
    priority_merge_label: Optional[str] = None
    # never merge a PR. This can be used with merge.update_branch_immediately to
    # automatically update a PR without merging.
    do_not_merge: bool = False


class Update(BaseModel):
    # update PR whenever the PR is out of date with the base branch. PR will be
    # updated regardless of failing requirements for merge (e.g. failing status
    # checks, missing reviews, blacklist labels). Kodiak will only update the PR
    # if the automerge label is enabled or `update.require_automerge_label` is
    # false.
    always: bool = False
    require_automerge_label: bool = True
    autoupdate_label: Optional[str] = None
    # Do not update PRs created by a listed user.
    blacklist_usernames: List[str] = []  # deprecated for ignored_usernames
    ignored_usernames: List[str] = []


class Approve(BaseModel):
    # auto approve any PR created by a listed user.
    auto_approve_usernames: List[str] = []
    # auto approve any PR to which a listed label is assigned.
    auto_approve_labels: List[str] = []


class InvalidVersion(ValueError):
    pass


class V1(BaseModel):
    version: int
    # _internal_ value to require a certain github app ID to process this repo.
    # This is useful for development, so we can globally install the production
    # kodiak, but also run the development version on a specific repo. By
    # setting _app_id to the development github app ID, we can prevent the
    # production kodiak instance from interfering.
    app_id: Optional[str]
    merge: Merge = Merge()
    update: Update = Update()
    approve: Approve = Approve()
    # when added to a Pull Request Kodiak will be prevented from taking any action
    # (approvals, updates, merges, comments, labels). Kodiak will still set
    # status checks. A user should generally not need to change this label as it
    # should rarely be applied.
    disable_bot_label: str = "kodiak:disabled"

    @validator("version", pre=True, always=True)
    def correct_version(cls, v: int) -> int:  # noqa: N805
        if v != 1:
            raise InvalidVersion("Version must be `1`")
        return v

    @classmethod
    def parse_toml(
        cls, content: str
    ) -> Union[V1, toml.TomlDecodeError, ValidationError]:
        try:
            return cls.parse_obj(toml.loads(content))
        except (toml.TomlDecodeError, ValidationError) as e:
            return e
